src/ExtMsg.actor.cpp

```diff
@@ -694,13 +689,22 @@ struct ExtIndexInsert : ConcreteInsertOp<ExtIndexInsert> {
 		state Optional<bson::BSONObj> idObj;
 		state Reference<QueryContext> dcx;
 
+		state bson::BSONObjBuilder indexObjToBeWritten;
 		Optional<IdInfo> encodedIds = extractEncodedIds(self->indexObj);
 
 		if (!encodedIds.present()) {
-			encodedId = DataValue(bson::OID::gen()).encode_key_part();
-			valueEncodedId = encodedId;
-			idObj = Optional<bson::BSONObj>();
+			// No _id field, generate a new one.
+			state bson::OID generatedId = bson::OID::gen();
```

Comment:
Can't this code reuse `insertDocument`?

Reply:
You mean the part that ensures id present? Yeah I think we can.